community/kernel/src/main/java/org/neo4j/kernel/impl/api/BatchTransactionApplier.java

```diff
@@ -0,0 +1,163 @@
+     * <pre>
+     *     try ( TransactionApplier txApplier = batchTxApplier.startTx( txToApply )
+     *     {
+     *         // Apply the transaction
+     *         txToApply.transactionRepresentation().accept( txApplier );
+     *         // Or apply other commands
+     *         // txApplier.visit( command );
+     *     }
+     * </pre>
+     *
+     * @param transaction The transaction which this applier is going to apply. Once we don't have to validate index
+     * updates anymore, we can change this to simply be the transactionId
+     * @param lockGroup A lockGroup which can hold the locks that the transaction requires.
+     * @return a {@link TransactionApplier} which can apply this transaction and other commands to the store.
+     */
+    TransactionApplier startTx( TransactionToApply transaction, LockGroup lockGroup ) throws IOException;
+
+    /**
+     * This method is suitable for any work that needs to be done after a batch of transactions. Typically called
+     * implicitly at the end of a try-with-resources block.
+     *
+     * @throws Exception
+     */
+    @Override
+    void close() throws Exception;
+
+    /**
+     * This class wraps several {@link BatchTransactionApplier}s which will do their work sequentially. See also {@link
+     * TransactionApplierFacade} which is used to wrap the {@link #startTx(TransactionToApply)} and {@link
+     * #startTx(TransactionToApply, LockGroup)} methods.
+     */
+    class BatchTransactionApplierFacade implements BatchTransactionApplier
+    {
+
+        private final BatchTransactionApplier[] appliers;
+
+        public BatchTransactionApplierFacade( BatchTransactionApplier... appliers )
+        {
+            this.appliers = appliers;
+        }
+
+        @Override
+        public TransactionApplier startTx( TransactionToApply transaction ) throws IOException
+        {
+            ArrayList<TransactionApplier> txAppliers = new ArrayList<>();
+            for ( BatchTransactionApplier applier : appliers )
+            {
+                txAppliers.add( applier.startTx( transaction ) );
+            }
+            return new TransactionApplierFacade(
+                    txAppliers.toArray( new TransactionApplier[appliers.length] ) );
+        }
+
+        @Override
+        public TransactionApplier startTx( TransactionToApply transaction, LockGroup lockGroup ) throws IOException
+        {
+            ArrayList<TransactionApplier> txAppliers = new ArrayList<>();
+            for ( BatchTransactionApplier applier : appliers )
+            {
+                txAppliers.add( applier.startTx( transaction, lockGroup ) );
+            }
+            return new TransactionApplierFacade(
+                    txAppliers.toArray( new TransactionApplier[appliers.length] ) );
+        }
+
+        @Override
+        public void close() throws Exception
+        {
+            // Not sure why it is necessary to close them in reverse order
```

Comment:
Yes it would but neither I nor Mattias know why the previous implementation did it like this. It was never documented